execution_engine/src/shared/storage_costs.rs

```diff
@@ -27,8 +24,12 @@ impl StorageCosts {
     }
 
     /// Calculates gas cost for storing `bytes`.
+    ///
+    /// This method will never overflow
     pub fn calculate_gas_cost(&self, bytes: usize) -> Gas {
-        let value = U512::from(self.gas_per_byte) * U512::from(bytes);
+        let gas_per_byte = self.gas_per_byte as u64;
+        let bytes_size = bytes as u64;
+        let value = gas_per_byte.saturating_mul(bytes_size);
```

Comment:
I wonder whether this should return `Option<Gas>` now:
If it returns `Some(u64::MAX.into())`, the cost was _exactly_ the maximum limit, which is fine. If it returns `None`, the gas limit was exceeded. (Probably not really relevant in practice, but still…)

Reply:
I think you're right. Let's consider the current gas_limit is 0, and I paid `u64::MAX` in the payment code. Then somehow, I tried to overflow the gas counter by writing gigabytes of data (let's assume it's possible), then technically, storage will already write the data as we'd end up in this branch of `charge_gas`: https://github.com/casper-network/casper-node/blob/b84d3909d7d05d8b3865dff692dbbb9ea86f1c11/execution_engine/src/core/runtime_context/mod.rs#L762-L765. The system will return the gas limit on the next charged operation (even a wasm NOP), but the data is already on disk.

I think the same can apply for `HostFunction::calculate_gas_cost` where I'm saturating the calculation, rather than exit early with None if such calculation would overflow.

So I guess saturating gas counter in operations that can overflow now could be considered risky? Maybe it always was 🤔

Reply:
Addressed in 50f8d34e95d92e401c8caf04332f9a7a77803e62 where I'm doing `checked_*` ops on Gas when charging for host function, and when charging for storage.